internal/exec/util/passwd.go

```diff
@@ -25,72 +25,121 @@ import (
 	keys "github.com/coreos/update-ssh-keys/authorized_keys_d"
 )
 
-// CreateUser creates the user as described.
-func (u Util) CreateUser(c types.PasswdUser) error {
-	if c.Create == nil {
-		return nil
+// EnsureUser ensures that the user exists as described. If the user does not
+// yet exist, they will be created, otherwise the existing user will be
+// modified.
+func (u Util) EnsureUser(c types.PasswdUser) error {
+	exists, err := u.CheckIfUserExists(c)
+	if err != nil {
+		return err
+	}
+	if c.Create != nil {
+		cu := c.Create
+		c.Gecos = cu.Gecos
+		c.Groups = translateV2_1UsercreateGroupSliceToPasswdUserGroupSlice(cu.Groups)
+		c.HomeDir = cu.HomeDir
+		c.NoCreateHome = cu.NoCreateHome
+		c.NoLogInit = cu.NoLogInit
+		c.NoUserGroup = cu.NoUserGroup
+		c.PrimaryGroup = cu.PrimaryGroup
+		c.Shell = cu.Shell
+		c.System = cu.System
+		c.UID = cu.UID
+	}
+	return u.CreateUser(c, exists)
+}
+
+// golang--
+func translateV2_1UsercreateGroupSliceToPasswdUserGroupSlice(groups []types.UsercreateGroup) []types.PasswdUserGroup {
+	newGroups := make([]types.PasswdUserGroup, len(groups))
+	for i, g := range groups {
+		newGroups[i] = types.PasswdUserGroup(g)
 	}
+	return newGroups
+}
 
-	cu := c.Create
+func (u Util) CheckIfUserExists(c types.PasswdUser) (bool, error) {
+	err := u.LogCmd(exec.Command("chroot", u.DestDir, "id", c.Name),
+		"checking if user %q exists", c.Name)
+	if err != nil {
```

Comment:
This doesn't distinguish between `id` returning an error and failing to run `id`.